descabi-morello/descabi-morello.rst

```diff
@@ -0,0 +1,769 @@
+
+
+Types Varying by Data Model and Procedure Calling Standard
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. _C/C++ type variants by data model and PCS:
+
+.. class:: aapcs64-morello-desc-c-cpp-type-variants
+
+.. table:: C/C++ type variants by data model and PCS
+
+    +-----------------------------+-------------------------------------+------------------------------+
+    | C/C++ Type                  | Machine Type                        | Notes                        |
+    +=============================+=====================================+==============================+
+    | ``T *``                     | Data Capability                     | Any data type ``T``.         |
+    +-----------------------------+-------------------------------------+------------------------------+
+    | ``T (*F)()``                | Function Capability                 | Any function type ``F``.     |
+    +-----------------------------+-------------------------------------+------------------------------+
+    | ``T&``                      | Data Capability                     | C++ reference.               |
+    +-----------------------------+-------------------------------------+------------------------------+
+    | ``T * __capability``        | Function Capability                 | Any data type ``T``.         |
+    +-----------------------------+-------------------------------------+------------------------------+
+    | ``T (* __capability F)()``  | Function Capability                 | Any function type ``F``.     |
+    +-----------------------------+-------------------------------------+------------------------------+
+    | ``T& __capability``         | Data Capability                     | C++ reference.               |
+    +-----------------------------+-------------------------------------+------------------------------+
+    | ``&&T``                     | Code Capability                     | Any label T.                 |
+    +-----------------------------+-------------------------------------+------------------------------+
+
+Morello Descriptor ABI ELF Extensions
+=====================================
+
+
+Private data sections
+---------------------
+
+The following sections are now part of a ``PT_MORELLO_DESC`` segment with the ``p_type`` value 0x70001000:
+
+- ``.desc.data.rel.ro``
+
+- ``.got``
+
+- ``.data``
+
+- ``.bss``
+
+
+The value of the Capability Private Data Addressing Register (c28) points to the start of the ``PT_MORELLO_DESC``
+segment, is read-only and can be used to access the ``.desc.data.rel.ro`` and ``.got`` sections.
+
+The PT_MORELLO_DESC segment is similar to a TLS segment and can be moved to a different address by
+the runtime of dynamic linker.
+
+PLT sequence
+------------
+
+The PLT sequence for the Morello Descriptor ABI is:
+
+.. code-block:: text
+
+    adrp c16, :got:foo
```

Comment:
I think you mean adrdp? Otherwise this is PC-relative so code and data are 1:1 again.

Reply:
It ends up as adrdp when the linker resolves the relocation. It might be better to just say adrdp here though to avoid confusion.